source/user/lmbench3/doc/lat_mmap.8
.\" $Id: lat_mmap.8,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $
.TH LAT_MMAP 8 "$Date: 2006-11-23 11:25:34 $" "(c)1994 Larry McVoy" "LMBENCH"
.SH NAME
lat_mmap \- costs of mmapping and unmmapping varying file sizes
.SH SYNOPSIS
.B lat_mmap
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
.I size
.I file
.SH DESCRIPTION
.B lat_mmap
times how fast a mapping can be made and unmade.  This is useful because it
is a fundemental part of processes that use SunOS style shared libraries
(the libraries are mapped in at process start up time and unmapped at 
process exit).
.LP
The benchmark maps in and unmaps the first \fIsize\fP bytes of the file
repeatedly and reports the average time for one mapping/unmapping.  
.LP
The size
specification may end with ``k'' or ``m'' to mean
kilobytes (* 1024) or megabytes (* 1024 * 1024).
.SH OUTPUT
Output format is \f(CB"%0.2f %d\\n", megabytes, usecs\fP, i.e.,
.sp
.ft CB
8.00 1200
.ft
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.


source/History
Ralink AP SDK - History

Version 4.0.1.0
===============

- [IPv6] Avoid both eth2.2 and eth2 set same IPV6 address after receiving RA packet in gateway mode.
- [IPv6] Fix HNAT misbinding issue (RT6855/6)
- [IPv6] Fix ping fail(ping ::1) issue of loopback interface when we try to enable and disable disable_ipv6 parameter in /proc directory (RT6855/6)
- [Switch] Fix legacy switch(RT3052/RT3352/RT5350) settings vlan 12345 error
- [PPPoE] Kernel panic after PAP fail (RT6855/6)
- [PPTP] Kernel panic after PPTP peer terminated the link (RT6855/6)
- [SPI Flash] There are some noise if we try to program flash while I2S is playing music (RT3052/RT3883/RT3352/RT5350)
- [GPIO] Fix GPIO interrupt handler cannot be trigged correctly (RT6855/6)
- [GUI] Fix WPS bug for test plan
- [NET] Recover irqs_disabled() check under SKB Recycle to fix warning messages, but it decreases RT6855/6's maximum peek throughput
- [Uboot] RT6855/6 does watchdog reset at cold start to fix timer interrupt issue

Version 4.0.0.0
===============

PLATFORM:

- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)

NEW FUNCTIONS:

- [IPv6 ready logo Phase II] RT6855/RT6856
- [IPv6] support IPv6 Rapid Deployment
- [IPv6] support IPv6 DS-Lite
- [GUI] support two giga phy port
- [GUI] display AP Client site_survey

IMPROVEMENT:

- [IPv6] support ipv6 tftp server
- [PPTP/L2TP] performace tunning at Linux 2.6.21 (PPTP improve 25%, L2TP improve 10% compare to SDK 3.6)
- [SAMBA] significant performance fine tune on full loading scenario (dual-band concurrent AP with USB access)
- [NAND Flash] config page size and address cycle based on chip mode
- [I2S] Add MMAP for user/kernel
- [GUI] support WPS 2.0 

Bug fixed:

- [Kernel NVRAM] memory leakage / initiate failed / free cache / wirte data exceed 1024
- [RAETH] fix STP drop issue on RT3052 chipst ONLY
- [RAETH] fix switch abnormal after ifconfig down&up
- [RAETH] fix PDMA hang issue under heavy load test
- [RAETH] fix TSO bug if USB storage is enabled
- [USB device] system will hang whiling loading gadget driver. Turning off power-saving on both port0 and port1
- [USB device] force host to do re-enumeration to solve "unrecognize device" when re-plug-in in windows platform.
- [IPv6] kernel panic bug if enable ipv6 privacy extension or router prereference
- [IPv6] fix NS packet loss for ipv6 DAD
- [GUI] flush conntrack table while restart/deleting rules
- [NAND Flash] fix nand flash not work for MTD_ANY mode
- [IGMPPROXY] fixed the overflow FATAL bug while handling bit47=1 mac address
- [GPIO] fixed RT3352/RT5350 extended GPIO pins mapping



source/user/lmbench3/src/lat_mmap.c
/*
 * lat_mmap.c - time how fast a mapping can be made and broken down
 *
 * Usage: mmap [-r] [-C] [-P <parallelism>] [-W <warmup>] [-N <repetitions>] size file
 *
 * XXX - If an implementation did lazy address space mapping, this test
 * will make that system look very good.  I haven't heard of such a system.
 *
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: lat_mmap.c,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $\n";

#include "bench.h"

#define	PSIZE	(16<<10)
#define	N	10
#define	STRIDE	(10*PSIZE)
#define	MINSIZE	(STRIDE*2)

#define	CHK(x)	if ((x) == -1) { perror("x"); exit(1); }


typedef struct _state {
	size_t	size;
	int	fd;
	int	random;
	int	clone;
	char	*name;
} state_t;

void	init(iter_t iterations, void *cookie);
void	cleanup(iter_t iterations, void *cookie);
void	domapping(iter_t iterations, void * cookie);

int
main(int ac, char **av)
{
	state_t state;
	int	parallel = 1;
	int	warmup = 0;
	int	repetitions = TRIES;
	char	buf[256];
	int	c;
	char	*usage = "[-r] [-C] [-P <parallelism>] [-W <warmup>] [-N <repetitions>] size file\n";
	

	state.random = 0;
	state.clone = 0;
	while (( c = getopt(ac, av, "rP:W:N:C")) != EOF) {
		switch(c) {
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0)
				lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		case 'r':
			state.random = 1;
			break;
		case 'C':
			state.clone = 1;
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}

	if (optind + 2 != ac) {
		lmbench_usage(ac, av, usage);
	}

	state.size = bytes(av[optind]);
	if (state.size < MINSIZE) {
		return (1);
	}
	state.name = av[optind+1];

	benchmp(init, domapping, cleanup, 0, parallel, 
		warmup, repetitions, &state);

	if (gettime() > 0) {
		micromb(state.size, get_n());
	}
	return (0);
}

void
init(iter_t iterations, void* cookie)
{
	state_t *state = (state_t *) cookie;
	
	if (iterations) return;

	if (state->clone) {
		char buf[128];
		char* s;

		/* copy original file into a process-specific one */
		sprintf(buf, "%d", (int)getpid());
		s = (char*)malloc(strlen(state->name) + strlen(buf) + 1);
		sprintf(s, "%s%d", state->name, (int)getpid());
		if (cp(state->name, s, S_IREAD|S_IWRITE) < 0) {
			perror("Could not copy file");
			unlink(s);
			exit(1);
		}
		state->name = s;
	}
	CHK(state->fd = open(state->name, O_RDWR));
	if (state->clone) unlink(state->name);
	if (lseek(state->fd, 0, SEEK_END) < state->size) {
		fprintf(stderr, "Input file too small\n");
		exit(1);
	}
}

void
cleanup(iter_t iterations, void* cookie)
{
	state_t *state = (state_t *) cookie;

	if (iterations) return;

	close(state->fd);
}

/*
 * This alg due to Linus.  The goal is to have both sparse and full
 * mappings reported.
 */
void
domapping(iter_t iterations, void *cookie)
{
	state_t *state = (state_t *) cookie;
	register int fd = state->fd;
	register size_t size = state->size;
	register int random = state->random;
	register char	*p, *where, *end;
	register char	c = size & 0xff;

	while (iterations-- > 0) {

#ifdef	MAP_FILE
		where = mmap(0, size, PROT_READ|PROT_WRITE, MAP_FILE|MAP_SHARED, fd, 0);
#else
		where = mmap(0, size, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
#endif
		if ((long)where == -1) {
			perror("mmap");
			exit(1);
		}
		if (random) {
			end = where + size;
			for (p = where; p < end; p += STRIDE) {
				*p = c;
			}
		} else {
			end = where + (size / N);
			for (p = where; p < end; p += PSIZE) {
				*p = c;
			}
		}
		munmap(where, size);
	}
}
